File: scripts/migrate_themed_styles.py
#!/usr/bin/env python3
"""Migrate static colors in StyleSheet.create to useThemedStyles."""
import os
import re
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent

SKIP = {
    "src/constants/theme/colors.ts",
    "src/constants/theme/palettes.ts",
    "src/constants/theme/types.ts",
    "src/components/ui/Text.tsx",
    "src/components/ui/Button.tsx",
    "src/components/ui/Input.tsx",
    "src/components/ui/TextArea.tsx",
    "src/components/layout/Screen.tsx",
    "src/components/layout/PageHeader.tsx",
    "src/features/settings/screens/SettingsHomeScreen.tsx",
    "src/features/settings/components/AppearanceSection.tsx",
    "src/features/cv-builder/constants/cv-docs-theme.ts",
    "src/features/mentorship/constants/theme.ts",
}


def rel(path: Path) -> str:
    return path.relative_to(ROOT).as_posix()


def walk_tsx(root: Path):
    for dirpath, dirnames, filenames in os.walk(root):
        dirnames[:] = [d for d in dirnames if d not in ("node_modules", ".git")]
        for name in filenames:
            if name.endswith((".tsx", ".ts")):
                yield Path(dirpath) / name


def migrate(content: str) -> str | None:
    if "colors." not in content:
        return None
    if "StyleSheet.create" not in content:
        return None
    if not re.search(r"from ['\"]@/constants/theme['\"]", content):
        return None
    if "useThemedStyles" in content:
        return None

    result = content

    def strip_colors(m):
        parts = [p.strip() for p in m.group(1).split(",") if p.strip() and p.strip() != "colors"]
        if not parts:
            return ""
        return f"import {{ {', '.join(parts)} }} from '@/constants/theme';"

    result = re.sub(
        r"import\s*\{([^}]+)\}\s*from\s*['\"]@/constants/theme['\"];",
        strip_colors,
        result,
        count=1,
    )

    if "from '@/hooks/useThemedStyles'" not in result:
        first_import = re.search(r"^import .+$", result, re.M)
        if first_import:
            insert = first_import.end() + 1
            result = (
                result[:insert]
                + "import type { ColorScheme } from '@/constants/theme/types';\n"
                + "import { useThemedStyles } from '@/hooks/useThemedStyles';\n"
                + result[insert:]
            )

    result = re.sub(
        r"const\s+styles\s*=\s*StyleSheet\.create\(\{",
        "function createStyles(colors: ColorScheme) {\n  return StyleSheet.create({",
        result,
        count=1,
    )

    if "function createStyles" not in result:
        return None

    idx = result.find("function createStyles")
    start = result.find("StyleSheet.create({", idx)
    depth = 0
    end = -1
    for i in range(start, len(result)):
        ch = result[i]
        if ch == "{":
            depth += 1
        elif ch == "}":
            depth -= 1
            if depth == 0:
                end = i
                break
    if end == -1:
        return None

    after = result[end : end + 3]
    if not after.startswith("});"):
        return None
    result = result[: end + 3] + "\n}" + result[end + 3 :]

    fns = []
    for m in re.finditer(r"export\s+function\s+\w+\s*\([^)]*\)\s*\{", result):
        fns.append(m.end())
    for m in re.finditer(r"export\s+default\s+function\s+\w*\s*\([^)]*\)\s*\{", result):
        fns.append(m.end())

    if not fns:
        return None

    for pos in sorted(fns, reverse=True):
        snippet = result[pos : pos + 200]
        if "useThemedStyles(createStyles)" in snippet:
            continue
        result = (
            result[:pos]
            + "\n  const styles = useThemedStyles(createStyles);"
            + result[pos:]
        )

    return result


def main():
    migrated = 0
    for base in (ROOT / "src", ROOT / "app"):
        if not base.exists():
            continue
        for path in walk_tsx(base):
            r = rel(path)
            if r in SKIP:
                continue
            text = path.read_text(encoding="utf-8")
            next_text = migrate(text)
            if next_text and next_text != text:
                path.write_text(next_text, encoding="utf-8")
                migrated += 1
                print(f"Migrated: {r}")
    print(f"\nDone. Migrated {migrated} files.")


if __name__ == "__main__":
    main()

File: scripts/migrate-themed-styles.mjs
/**
 * Migrates files using static `colors` in StyleSheet.create to useThemedStyles.
 * Run: node scripts/migrate-themed-styles.mjs
 */
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.join(__dirname, '..');

const SKIP_FILES = new Set([
  'src/constants/theme/colors.ts',
  'src/constants/theme/palettes.ts',
  'src/constants/theme/types.ts',
  'src/components/ui/Text.tsx',
  'src/components/ui/Button.tsx',
  'src/components/ui/Input.tsx',
  'src/components/ui/TextArea.tsx',
  'src/components/layout/Screen.tsx',
  'src/components/layout/PageHeader.tsx',
  'src/features/settings/screens/SettingsHomeScreen.tsx',
  'src/features/settings/components/AppearanceSection.tsx',
  'src/features/help/components/SupportContactHint.tsx',
  'src/features/cv-builder/constants/cv-docs-theme.ts',
  'src/features/mentorship/constants/theme.ts',
]);

function walk(dir, files = []) {
  for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
    const full = path.join(dir, entry.name);
    if (entry.isDirectory()) {
      if (entry.name === 'node_modules' || entry.name === '.git') continue;
      walk(full, files);
    } else if (/\.(tsx|ts)$/.test(entry.name)) {
      files.push(full);
    }
  }
  return files;
}

function rel(p) {
  return path.relative(ROOT, p).replace(/\\/g, '/');
}

function migrate(content, filePath) {
  if (!content.includes('colors.')) return null;
  if (!content.includes('StyleSheet.create')) return null;
  if (!content.match(/from ['"]@\/constants\/theme['"]/)) return null;
  if (content.includes('useThemedStyles')) return null;

  let result = content;

  // Update theme import: remove colors, keep others
  result = result.replace(
    /import\s*\{([^}]+)\}\s*from\s*['"]@\/constants\/theme['"];/,
    (match, imports) => {
      const parts = imports
        .split(',')
        .map((s) => s.trim())
        .filter((s) => s && s !== 'colors');
      if (parts.length === 0) return '';
      return `import { ${parts.join(', ')} } from '@/constants/theme';`;
    },
  );

  // Add new imports if not present
  if (!result.includes("from '@/hooks/useThemedStyles'")) {
    const importAnchor = result.match(/^import .+$/m);
    if (importAnchor) {
      const pos = result.indexOf(importAnchor[0]);
      const lastImportIdx = result.lastIndexOf('\nimport ', pos + 5000);
      const insertAt = result.indexOf('\n', lastImportIdx >= 0 ? lastImportIdx : pos) + 1;
      result =
        result.slice(0, insertAt) +
        "import type { ColorScheme } from '@/constants/theme/types';\n" +
        "import { useThemedStyles } from '@/hooks/useThemedStyles';\n" +
        result.slice(insertAt);
    }
  }

  // Convert const styles = StyleSheet.create to function createStyles
  result = result.replace(
    /const\s+styles\s*=\s*StyleSheet\.create\(\{/,
    'function createStyles(colors: ColorScheme) {\n  return StyleSheet.create({',
  );

  // Close createStyles function - find the closing }); of styles
  const createIdx = result.indexOf('function createStyles');
  if (createIdx === -1) return null;

  // Find matching close for StyleSheet.create
  const start = result.indexOf('StyleSheet.create({', createIdx);
  let depth = 0;
  let end = -1;
  for (let i = start; i < result.length; i++) {
    if (result[i] === '{') depth++;
    if (result[i] === '}') {
      depth--;
      if (depth === 0) {
        end = i;
        break;
      }
    }
  }
  if (end === -1) return null;

  // After `});` add closing brace for function
  const afterClose = result.slice(end);
  const closeMatch = afterClose.match(/^\}\);/);
  if (!closeMatch) return null;
  const insertPos = end + closeMatch[0].length;
  result = result.slice(0, insertPos) + '\n}' + result.slice(insertPos);

  // Add useThemedStyles to exported function components
  const fnPattern = /export\s+function\s+(\w+)\s*\([^)]*\)\s*\{/g;
  let fnMatch;
  const fns = [];
  while ((fnMatch = fnPattern.exec(result)) !== null) {
    fns.push({ name: fnMatch[1], index: fnMatch.index + fnMatch[0].length });
  }

  // Also default export functions
  const defaultFnPattern = /export\s+default\s+function\s+(\w+)?\s*\([^)]*\)\s*\{/g;
  while ((fnMatch = defaultFnPattern.exec(result)) !== null) {
    fns.push({ name: fnMatch[1] || 'default', index: fnMatch.index + fnMatch[0].length });
  }

  if (fns.length === 0) return null;

  // Insert useThemedStyles in each function (reverse order to preserve indices)
  fns.sort((a, b) => b.index - a.index);
  for (const fn of fns) {
    const bodyStart = fn.index;
    const snippet = result.slice(bodyStart, bodyStart + 200);
    if (snippet.includes('useThemedStyles(createStyles)')) continue;
    result =
      result.slice(0, bodyStart) +
      '\n  const styles = useThemedStyles(createStyles);' +
      result.slice(bodyStart);
  }

  return result;
}

const dirs = [path.join(ROOT, 'src'), path.join(ROOT, 'app')];
let migrated = 0;
let skipped = 0;

for (const dir of dirs) {
  for (const file of walk(dir)) {
    const r = rel(file);
    if (SKIP_FILES.has(r)) {
      skipped++;
      continue;
    }
    const content = fs.readFileSync(file, 'utf8');
    const next = migrate(content, file);
    if (next && next !== content) {
      fs.writeFileSync(file, next);
      migrated++;
      console.log('Migrated:', r);
    }
  }
}

console.log(`\nDone. Migrated ${migrated} files, skipped ${skipped} reserved files.`);
